stdlib: Improve handing of errors in Atlas request failures

BobbyRBruce opened the pull request:
Now:

* The Atlas Client will attempt a connection 4 times, using an exponential backoff approach between attempts.
* When a failure does arise a rich output is given so problems can be easily diagnosed.

Addresses: #340

BobbyRBruce committed:
stdlib: Improve handing of errors in Atlas request failures

Now:

* The Atlas Client will attempt a connection 4 times, using an
  exponential backoff approach between attempts.
* When a failure does arise a rich output is given so problems can be
  easily diagnosed.

Change-Id: I3df332277c33a040c0ed734b9f3e28f38606af44

BobbyRBruce commented:
@Harshil2107 : I was looking over #392 and had a few suggestions. This then spiraled into me just doing it. Open for feedback here. I haven't tested this yet (so pseudo-drafty), but I think the design idea is sound.

BobbyRBruce commented:
I don't know what's going on, but the tests are failing with the 404 error, so at least we can see what the error would look like to the user:

```sh
warn: `Resource` has been deprecated. Please use the `obtain_resource` function instead.
warn: Attempt 1 of Atlas HTTP Request failed.
Purpose of Request: Get Access Token with API key.

Failed with Exception:
HTTP Error 404: Not Found

Retrying after 2 seconds...
warn: Attempt 2 of Atlas HTTP Request failed.
Purpose of Request: Get Access Token with API key.

Failed with Exception:
HTTP Error 404: Not Found

Retrying after 4 seconds...
warn: Attempt 3 of Atlas HTTP Request failed.
Purpose of Request: Get Access Token with API key.

Failed with Exception:
HTTP Error 404: Not Found

Retrying after 8 seconds...
Exception: Error getting resources from client gem5-resources

At:
  src/python/gem5/resources/client_api/client_wrapper.py(117): get_all_resources_by_id
  src/python/gem5/resources/client_api/client_wrapper.py(150): get_resource_json_obj_from_client
  src/python/gem5/resources/client.py(152): get_resource_json_obj
  src/python/gem5/resources/resource.py(736): obtain_resource
  src/python/gem5/resources/resource.py(975): Resource
  /__w/gem5/gem5/tests/gem5/m5_util/configs/simple_binary_run.py(92): <module>
  src/python/m5/main.py(642): main
```

This somehow isn't giving the errors information from the `AtlasClient` like i thought they'd be. This is a bug.

BobbyRBruce commented:
Think i've fixed the issue partially in the latest push... we'll see.

Harshil2107 commented on src/python/gem5/resources/client_api/atlasclient.py:
```diff
@@ -47,22 +77,70 @@ def __init__(self, config: Dict[str, str]):
+        self,
+        url: str,
+        data_json: Dict[str, Any],
+        headers: dict[str, str],
```
Shouldn't this be `Dict[str, str]`?

Harshil2107 commented on src/python/gem5/resources/client_api/atlasclient.py:
```diff
@@ -47,22 +77,70 @@ def __init__(self, config: Dict[str, str]):
+                    raise AtlasClientHttpJsonRequestError(
+                        client=self,
+                        data=data_json,
+                        headers=headers,
```
`header` is being passed as a parameter to `AtlasClientHttpJsonRequestError` but it does not have a parameter named `header`.

Harshil2107 committed:
stdlib, tests: Fixed bugs and tests

- Fixed bugs rekated to retrying on request faliure.
- Updated the pyunit tests.

Change-Id: Ia484690267bf27018488324f3408f7e47c59bef3

powerjg reviewed (approved):
lgtm
